jetty-jmx/src/main/java/org/eclipse/jetty/jmx/MBeanContainer.java

```diff
@@ -122,6 +114,215 @@ public String getDomain()
+            }
+            catch (ClassNotFoundException e)
+            {
+                // Not found, so if not the same as the thread context loader, try that.
+                if (Thread.currentThread().getContextClassLoader() == oClass.getClassLoader())
+                    throw e;
+                LOG.ignore(e);
+                mClass = Loader.loadClass(oClass, mName);
+            }
+
+            constructor =  (ModelMBean.class.isAssignableFrom(mClass))
+                ?mClass.getDeclaredConstructor()
+                :mClass.getConstructor(OBJ_ARG);
+        }
+        catch (ClassNotFoundException | NoSuchMethodException e)
+        {
+            LOG.debug(e.toString());
+            LOG.ignore(e);
+        }
+
+        if (constructor==null)
+            constructor = findMBeanFor(oClass.getSuperclass());
+
+        // Can we cache the result?
+        ClassLoader ourLoader = this.getClass().getClassLoader();
+        if (constructor!=null && (_cacheOtherClassLoaders || oClass.getClassLoader()==ourLoader && constructor.getDeclaringClass().getClassLoader()==ourLoader))
+        {
+            Constructor<?> c = _mbeanFor.putIfAbsent(oClass, constructor);
+            if (c!=null)
+                constructor = c;
+        }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("findMBeanFor {} => {}", oClass, constructor);
+
+        return constructor;
+    }
+
+
+    MBeanInfo getMBeanInfo(ObjectMBean bean)
+    {
+        Object managed = bean.getManagedObject();
+        Class[] classes;
+        if (managed.getClass()!=bean.getClass())
+            classes = new Class[] {managed.getClass(), bean.getClass()};
+        else
+            classes = new Class[] {managed.getClass()};
+
+        String desc = null;
+        List<MBeanAttributeInfo> attributes = new ArrayList<>();
+        List<MBeanOperationInfo> operations = new ArrayList<>();
+
+        for (Class c : classes)
+        {
+            MetaData metadata = findMetaData(c);
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("MBean meta {} -> {}", bean, metadata);
+
+            if (desc==null && metadata._managedObject!=null)
+                desc = metadata._managedObject.value();
+
+            for (int i = 0; i < metadata._attributes.size(); i++)
+            {
+                MBeanAttributeInfo info = bean.defineAttribute(metadata._attributes.get(i), metadata._getters.get(i));
+                if (info != null)
+                    attributes.add(info);
+            }
+
+            for (int i = 0; i < metadata._operations.size(); i++)
+            {
+                MBeanOperationInfo info = bean.defineOperation(metadata._operations.get(i), metadata._methods.get(i));
+                if (info != null)
+                    operations.add(info);
+            }
+        }
+        return new MBeanInfo(managed.getClass().getName(),
```

Comment:
We are recreating the `MBeanInfo` objects for different instances. Because `MBeanInfo` is a descriptor of the _class_, we can share it among instances (e.g. among multiple `HTTP2Session` instances). Class `MetaData` should hold `MBean*Info` so we don't build it every time.

Reply:
I was trying to avoid looking inside the define methods... but now you've volunteered to do so... yes :)